an optional arg of type `Bar[K] | null` can not receive `null`

**TypeScript Version:**  2.2.0-dev.20170130 (npm version) / master

**Code**

```ts
interface Bar {  bar: number;  foo: object | null;  }

let a = {
  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  {
    console.log(a, b);
  }
};
a.test("bar", null);
```

**Expected behavior:**
`tsc --strictNullChecks` compiles it without errors.

**Actual behavior:**
It reports: 
```
error TS2345: Argument of type 'null' is not assignable to parameter of type 'number | undefin
ed'.
```

**Notes**
This case only occurs if at least one member of `Bar` can be null and `--strictNullChecks` is enabled.

I think the checker should not merge `null` and `undefined` in optional union-type args.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -8540,7 +8540,7 @@ namespace ts {
             if (flags & TypeFlags.Void) types.push(voidType);
             if (flags & TypeFlags.Undefined) types.push(undefinedType);
             if (flags & TypeFlags.Null) types.push(nullType);
-            return getUnionType(types, /*subtypeReduction*/ true);
+            return getUnionType(types);
         }
 
         function removeDefinitelyFalsyTypes(type: Type): Type {

--- a/tests/baselines/reference/optionalParameterRetainsNull.js
+++ b/tests/baselines/reference/optionalParameterRetainsNull.js
@@ -1,0 +1,14 @@
+//// [optionalParameterRetainsNull.ts]
+interface Bar {  bar: number; foo: object | null;  }
+
+let a = {
+  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  { }
+};
+a.test("bar", null); // ok, null is assignable to number | null | undefined
+
+
+//// [optionalParameterRetainsNull.js]
+var a = {
+    test: function (a, b) { }
+};
+a.test("bar", null); // ok, null is assignable to number | null | undefined

--- a/tests/baselines/reference/optionalParameterRetainsNull.symbols
+++ b/tests/baselines/reference/optionalParameterRetainsNull.symbols
@@ -1,0 +1,25 @@
+=== tests/cases/compiler/optionalParameterRetainsNull.ts ===
+interface Bar {  bar: number; foo: object | null;  }
+>Bar : Symbol(Bar, Decl(optionalParameterRetainsNull.ts, 0, 0))
+>bar : Symbol(Bar.bar, Decl(optionalParameterRetainsNull.ts, 0, 15))
+>foo : Symbol(Bar.foo, Decl(optionalParameterRetainsNull.ts, 0, 29))
+
+let a = {
+>a : Symbol(a, Decl(optionalParameterRetainsNull.ts, 2, 3))
+
+  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  { }
+>test : Symbol(test, Decl(optionalParameterRetainsNull.ts, 2, 9))
+>K : Symbol(K, Decl(optionalParameterRetainsNull.ts, 3, 7))
+>Bar : Symbol(Bar, Decl(optionalParameterRetainsNull.ts, 0, 0))
+>a : Symbol(a, Decl(optionalParameterRetainsNull.ts, 3, 29))
+>K : Symbol(K, Decl(optionalParameterRetainsNull.ts, 3, 7))
+>b : Symbol(b, Decl(optionalParameterRetainsNull.ts, 3, 34))
+>Bar : Symbol(Bar, Decl(optionalParameterRetainsNull.ts, 0, 0))
+>K : Symbol(K, Decl(optionalParameterRetainsNull.ts, 3, 7))
+
+};
+a.test("bar", null); // ok, null is assignable to number | null | undefined
+>a.test : Symbol(test, Decl(optionalParameterRetainsNull.ts, 2, 9))
+>a : Symbol(a, Decl(optionalParameterRetainsNull.ts, 2, 3))
+>test : Symbol(test, Decl(optionalParameterRetainsNull.ts, 2, 9))
+

--- a/tests/baselines/reference/optionalParameterRetainsNull.types
+++ b/tests/baselines/reference/optionalParameterRetainsNull.types
@@ -1,0 +1,31 @@
+=== tests/cases/compiler/optionalParameterRetainsNull.ts ===
+interface Bar {  bar: number; foo: object | null;  }
+>Bar : Bar
+>bar : number
+>foo : object | null
+>null : null
+
+let a = {
+>a : { test<K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined): void; }
+>{  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  { }} : { test<K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined): void; }
+
+  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  { }
+>test : <K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined) => void
+>K : K
+>Bar : Bar
+>a : K
+>K : K
+>b : Bar[K] | null | undefined
+>Bar : Bar
+>K : K
+>null : null
+
+};
+a.test("bar", null); // ok, null is assignable to number | null | undefined
+>a.test("bar", null) : void
+>a.test : <K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined) => void
+>a : { test<K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined): void; }
+>test : <K extends "bar" | "foo">(a: K, b?: Bar[K] | null | undefined) => void
+>"bar" : "bar"
+>null : null
+

--- a/tests/cases/compiler/optionalParameterRetainsNull.ts
+++ b/tests/cases/compiler/optionalParameterRetainsNull.ts
@@ -1,0 +1,7 @@
+// @strictNullChecks: true
+interface Bar {  bar: number; foo: object | null;  }
+
+let a = {
+  test<K extends keyof Bar> (a: K,  b?: Bar[K]  |  null)  { }
+};
+a.test("bar", null); // ok, null is assignable to number | null | undefined

